File: zxmap/src/main/java/com/lt/zxmap/view/MeaOptManager.java
package com.lt.zxmap.view;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.lt.zxmap.R;
import com.zxmap.zxmapsdk.annotations.IconFactory;
import com.zxmap.zxmapsdk.annotations.Marker;
import com.zxmap.zxmapsdk.annotations.MarkerViewOptions;
import com.zxmap.zxmapsdk.geojson.Feature;
import com.zxmap.zxmapsdk.geojson.LineString;
import com.zxmap.zxmapsdk.geojson.MultiLineString;
import com.zxmap.zxmapsdk.geojson.MultiPoint;
import com.zxmap.zxmapsdk.geojson.MultiPolygon;
import com.zxmap.zxmapsdk.geojson.Point;
import com.zxmap.zxmapsdk.geojson.Polygon;
import com.zxmap.zxmapsdk.geometry.LatLng;
import com.zxmap.zxmapsdk.maps.ZXMap;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Xiangb on 2019/9/27.
 * 功能：
 */
public class MeaOptManager {

    private Context context;
    private ZXMap zxMap;

    private List<OptBean> optUndoList = new ArrayList<>();//上一步的操作集合
    private List<OptBean> optRedoList = new ArrayList<>();//下一步的操作集合

    private MarkerViewOptions markerOptions;//标记点
    public LatLng popMenuLatlng;//弹出窗口时的定位点
    public int funcIndex = -1;//操作点

    private MeaListener meaListener;

    MeaOptManager(Context context, ZXMap zxMap, MeaListener meaListener) {
        this.context = context;
        this.zxMap = zxMap;
        this.meaListener = meaListener;
    }

    /**
     * 操作实体类
     */
    class OptBean {
        private static final int ADD = 1;
        private static final int MOVE = 2;
        private static final int INSERT = 3;
        private static final int DELETE = 4;
        private static final int TRANS = 5;
        private static final int MultiAdd = 6;

        private int optType;

        private LatLng pointAdd;
        private LatLng pointMoveFrom;
        private LatLng pointMoveTo;
        private LatLng pointInsert;
        private LatLng pointDelete;
        private LatLng pointTransStart;
        private LatLng pointTransEnd;
        private List<LatLng> pointsMulti = new ArrayList<>();

        private int moveIndex;
        private int insertIndex;
        private int deleteIndex;
    }

    //添加操作
    public void optAdd(LatLng latlng) {
        OptBean bean = new OptBean();
        bean.optType = OptBean.ADD;
        bean.pointAdd = latlng;
        optUndoList.add(bean);
        doOpt(true, bean, false);
        //清空下一步操作序列
        optRedoList.clear();
        meaListener.onMapReAdd();
    }

    //移动操作
    public void optMove(int moveIndex, LatLng latlngFrom, LatLng latlngTo) {
        OptBean bean = new OptBean();
        bean.optType = OptBean.MOVE;
        bean.pointMoveFrom = latlngFrom;
        bean.pointMoveTo = latlngTo;
        bean.moveIndex = moveIndex;
        optUndoList.add(bean);
        //清空下一步操作序列
        optRedoList.clear();
        meaListener.onMapReAdd();
    }

    //插入操作
    public void optInsert(LatLng latlng, int index) {
        OptBean bean = new OptBean();
        bean.optType = OptBean.INSERT;
        bean.pointInsert = latlng;
        bean.insertIndex = index;
        optUndoList.add(bean);
        doOpt(true, bean, false);
        //清空下一步操作序列
        optRedoList.clear();
        meaListener.onMapReAdd();
    }

    //删除操作
    public void optDelete(LatLng latlng, int index) {
        OptBean bean = new OptBean();
        bean.optType = OptBean.DELETE;
        bean.pointDelete = latlng;
        bean.deleteIndex = index;
        optUndoList.add(bean);
        doOpt(true, bean, false);
        //清空下一步操作序列
        optRedoList.clear();
        meaListener.onMapReAdd();
    }

    //平移操作
    public void optTrans(LatLng transStart, LatLng transEnd) {
        OptBean bean = new OptBean();
        bean.optType = OptBean.TRANS;
        bean.pointTransStart = transStart;
        bean.pointTransEnd = transEnd;
        optUndoList.add(bean);
        //清空下一步操作序列
        optRedoList.clear();
        meaListener.onMapReAdd();
    }

    //多点添加
    public void optMultiAdd(List<LatLng> latLngs) {
        OptBean bean = new OptBean();
        bean.optType = OptBean.MultiAdd;
        bean.pointsMulti.addAll(latLngs);
        optUndoList.add(bean);
        //清空下一步操作序列
        doOpt(true, bean, false);
        optRedoList.clear();
        meaListener.onMapReAdd();
    }

    /**
     * 上一步
     *
     * @return 是否已为最后一步
     */
    public boolean postUndo() {
        meaListener.stopTranslate();
        if (optUndoList.size() > 0) {
            doOpt(false, optUndoList.get(optUndoList.size() - 1));
            return optUndoList.size() == 0;
        }
        return false;
    }

    /**
     * 下一步
     *
     * @return 是否已为最后一步
     */
    public boolean postRedo() {
        meaListener.stopTranslate();
        if (optRedoList.size() > 0) {
            doOpt(true, optRedoList.get(optRedoList.size() - 1));
            return optRedoList.size() == 0;
        }
        return false;
    }

    /**
     * 清除
     */
    public void clear() {
        optUndoList.clear();
        optRedoList.clear();
    }

    private void doOpt(boolean isRedo, OptBean bean) {
        doOpt(isRedo, bean, true);
    }

    /**
     * 执行操作
     *
     * @param isRedo 是否为下一步操作，是则正向执行，否则反向执行
     * @param bean
     * @param isPost 是否是通过postRedo或者postUndo执行的
     */
    private void doOpt(boolean isRedo, OptBean bean, boolean isPost) {
        switch (bean.optType) {
            case OptBean.ADD:
                if (isRedo) {
                    meaListener.getLatlngs().add(bean.pointAdd);
                } else {
                    MeaUtil.removeLast(meaListener.getLatlngs(), bean.pointAdd);
//                    meaListener.getLatlngs().remove(bean.pointAdd);
                }
                break;
            case OptBean.MOVE:
                if (isRedo) {
                    meaListener.getLatlngs().set(bean.moveIndex, bean.pointMoveTo);
                } else {
                    meaListener.getLatlngs().set(bean.moveIndex, bean.pointMoveFrom);
                }
                break;
            case OptBean.INSERT:
                if (isRedo) {
                    meaListener.getLatlngs().add(bean.insertIndex, bean.pointInsert);
                } else {
                    meaListener.getLatlngs().remove(bean.insertIndex);
                }
                break;
            case OptBean.DELETE:
                if (isRedo) {
                    meaListener.getLatlngs().remove(bean.deleteIndex);
                } else {
                    meaListener.getLatlngs().add(bean.deleteIndex, bean.pointDelete);
                }
                break;
            case OptBean.TRANS:
                double transLon = bean.pointTransEnd.getLongitude() - bean.pointTransStart.getLongitude();
                double transLat = bean.pointTransEnd.getLatitude() - bean.pointTransStart.getLatitude();
                if (isRedo) {
                    for (int i = 0; i < meaListener.getLatlngs().size(); i++) {
                        meaListener.getLatlngs().get(i).setLongitude(meaListener.getLatlngs().get(i).getLongitude() + transLon);
                        meaListener.getLatlngs().get(i).setLatitude(meaListener.getLatlngs().get(i).getLatitude() + transLat);
                    }
                } else {
                    for (int i = 0; i < meaListener.getLatlngs().size(); i++) {
                        meaListener.getLatlngs().get(i).setLongitude(meaListener.getLatlngs().get(i).getLongitude() - transLon);
                        meaListener.getLatlngs().get(i).setLatitude(meaListener.getLatlngs().get(i).getLatitude() - transLat);
                    }
                }
                break;
            case OptBean.MultiAdd:
                if (isRedo) {
                    meaListener.getLatlngs().addAll(bean.pointsMulti);
                } else {
                    meaListener.getLatlngs().clear();
                }
                break;
            default:
                break;
        }
        meaListener.updateSource(false);
        if (isPost) {
            if (isRedo) {
                optRedoList.remove(bean);
                optUndoList.add(bean);
            } else {
                optRedoList.add(bean);
                optUndoList.remove(bean);
            }
        }
    }


    /**
     * 执行长按事件-弹出操作菜单
     */
    public void showMeaMenu(final Feature clickFeature) {
        closeMarker();
        zxMap.setAllowConcurrentMultipleOpenInfoWindows(false);
        zxMap.setInfoWindowAdapter(new ZXMap.InfoWindowAdapter() {
            @Nullable
            @Override
            public View getInfoWindow(@NonNull Marker marker) {
                View view = LayoutInflater.from(context).inflate(R.layout.map_meamenu_layout, null, false);
                LinearLayout llMenu = view.findViewById(R.id.ll_mea_menu);
                List<String> funcString = new ArrayList<>();
                if (clickFeature.getGeometry() instanceof Point || clickFeature.getGeometry() instanceof MultiPoint) {
                    funcString.add("删除");
                } else if (clickFeature.getGeometry() instanceof LineString || clickFeature.getGeometry() instanceof MultiLineString) {
                    funcString.add("插入");
//                    funcString.add("裁切");
                } else if (clickFeature.getGeometry() instanceof Polygon || clickFeature.getGeometry() instanceof MultiPolygon) {
                    if (!meaListener.isTranslate()) {
                        funcString.add("平移");
//                    funcString.add("旋转");
//                    funcString.add("裁切");
                    } else {
                        funcString.add("结束平移");
                    }
                }
                if (funcString.size() > 0) {
                    LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, MeaUtil.dp2px(context, 30));
                    params.gravity = Gravity.CENTER;
                    for (int i = 0; i < funcString.size(); i++) {
                        TextView tv = new TextView(context);
                        tv.setText(funcString.get(i));
                        tv.setTextColor(Color.WHITE);
                        tv.setTextSize(12);
                        tv.setGravity(Gravity.CENTER);
                        tv.setMinWidth(MeaUtil.dp2px(context, 30));
                        tv.setLayoutParams(params);
                        llMenu.addView(tv);
                        tv.setTag(funcString.get(i));
                        tv.setOnClickListener(new MenuClickListener());
                        if (funcString.size() > 1 && i < funcString.size() - 1) {
                            View divider = new View(context);
                            divider.setLayoutParams(new LinearLayout.LayoutParams(MeaUtil.dp2px(context, 1), MeaUtil.dp2px(context, 20)));
                            divider.setBackgroundColor(Color.WHITE);
                            llMenu.addView(divider);
                        }
                    }
                }

                return view;
            }
        });
        if (popMenuLatlng != null) {
            markerOptions = new MarkerViewOptions()
                    .position(popMenuLatlng)
                    .icon(IconFactory.getInstance(context)
                            .fromBitmap(Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)))
                    .alpha(0.0f);
            zxMap.addMarker(markerOptions);
            zxMap.selectMarker(markerOptions.getMarker());
        }
    }

    //关闭弹框
    public boolean closeMarker() {
        if (markerOptions != null && markerOptions.getMarker() != null) {
            markerOptions.getMarker().hideInfoWindow();
            markerOptions.getMarker().remove();
            markerOptions = null;
            zxMap.removeAnnotations();
            return true;
        } else {
            return false;
        }
    }

    //菜单点击事件
    private class MenuClickListener implements View.OnClickListener {

        @Override
        public void onClick(View v) {
            switch (v.getTag().toString()) {
                case "删除":
                    if (funcIndex != -1) {
                        optDelete(popMenuLatlng, funcIndex);
                    }
                    break;
                case "插入":
                    if (funcIndex != -1) {
                        optInsert(popMenuLatlng, funcIndex);
                    }
                    break;
                case "裁切":

                    break;
                case "平移":
                    meaListener.startTranslate();
                    break;
                case "结束平移":
                    meaListener.stopTranslate();
                    break;
                case "旋转":

                    break;
            }
            closeMarker();
        }
    }

}

File: module_map/src/main/java/com/zx/module_map/view/measure/MeaOptManager.kt
package com.gt.module_map.view.measure

import android.content.Context
import android.graphics.Color
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.esri.arcgisruntime.geometry.Point
import com.esri.arcgisruntime.geometry.Polygon
import com.esri.arcgisruntime.geometry.Polyline
import com.esri.arcgisruntime.mapping.view.Callout
import com.esri.arcgisruntime.mapping.view.Graphic
import com.gt.module_map.R
import com.zx.zxutils.util.ZXLogUtil

/**
 * 绘制操作管理器
 */
class MeaOptManager(private var context: Context, private var meaListener: MeaListener) {

    companion object {
        private const val MENU_ADD = "添加"
        private const val MENU_DELETE = "删除"
        private const val MENU_INSERT = "插入"
        private const val MENU_CUT = "裁切"
        private const val MENU_TRANS = "平移"
        private const val MENU_ROTATE = "旋转"
        private const val MENU_STOPTRANS = "结束平移"
    }

    /**
     * 操作实体类
     */
    data class OptBean(
        var optType: OptType? = null,

        var pointAdd: Point? = null,
        var pointMoveFrom: Point? = null,
        var pointMoveTo: Point? = null,
        var pointInsert: Point? = null,
        var pointDelete: Point? = null,
        var pointTransStart: Point? = null,
        var pointTransEnd: Point? = null,

        var addIndex: Int? = null,
        var moveIndex: Int? = null,
        var insertIndex: Int? = null,
        var deleteIndex: Int? = null
    ) {
        enum class OptType {
            ADD,
            MOVE,
            INSERT,
            DELETE,
            TRNAS
        }
    }

    private val optUndoList = arrayListOf<OptBean>()//上一步操作集合
    private val optRedoList = arrayListOf<OptBean>()//下一步操作集合

    var popMenuPoint: Point? = null//菜单弹出点位置
    var funcIndex = -1 //操作点


    /**
     * 加入添加操作
     */
    fun optAdd(point: Point) {
        val bean = OptBean().apply {
            optType = OptBean.OptType.ADD
            pointAdd = point
            addIndex = meaListener.getPointList().size
        }
        optUndoList.add(bean)
        doOpt(bean)
        //清空下一步操作序列
        optRedoList.clear()
        meaListener.onMapReAdd()
    }

    /**
     * 加入移动操作
     */
    fun optMove(moveIndex: Int, pointFrom: Point, pointTo: Point) {
        val bean = OptBean().apply {
            optType = OptBean.OptType.MOVE
            pointMoveFrom = pointFrom
            pointMoveTo = pointTo
            this.moveIndex = moveIndex
        }
        optUndoList.add(bean)
        //清空下一步操作序列
        optRedoList.clear()
        meaListener.onMapReAdd()
    }

    /**
     * 加入插入操作
     */
    fun optInsert(point: Point, index: Int) {
        val bean = OptBean().apply {
            optType = OptBean.OptType.INSERT
            pointInsert = point
            insertIndex = index
        }
        optUndoList.add(bean)
        doOpt(bean)
        //清空下一步操作序列
        optRedoList.clear()
        meaListener.onMapReAdd()
    }

    /**
     * 加入删除操作
     */
    fun optDelete(point: Point, index: Int) {
        val bean = OptBean().apply {
            optType = OptBean.OptType.DELETE
            pointDelete = point
            deleteIndex = index
        }
        optUndoList.add(bean)
        doOpt(bean)
        //清空下一步操作序列
        optRedoList.clear()
        meaListener.onMapReAdd()
    }

    /**
     * 加入平移操作
     */
    fun optTrans(pointStart: Point, pointEnd: Point) {
        val bean = OptBean().apply {
            optType = OptBean.OptType.TRNAS
            pointTransStart = pointStart
            pointTransEnd = pointEnd
        }
        optUndoList.add(bean)
        //清空下一步操作序列
        optRedoList.clear()
    }

    /**
     * 执行上一步操作
     * @return 是否已为最后一步
     */
    fun postUndo(): Boolean {
        meaListener.stopTranslate()
        if (optUndoList.isNotEmpty()) {
            doOpt(optUndoList.last(), false, true)
            return optUndoList.isEmpty()
        }
        return false
    }

    /**
     * 执行下一步操作
     * @return 是否已为最后一步
     */
    fun postRedo(): Boolean {
        meaListener.stopTranslate()
        if (optRedoList.isNotEmpty()) {
            doOpt(optRedoList.last(), true, true)
            return optRedoList.isEmpty()
        }
        return false
    }

    /**
     * 清除
     */
    fun clear() {
        optUndoList.clear()
        optRedoList.clear()
    }

    /**
     * 执行操作
     * @param bean
     * @param idRedo 是否为下一步操作，是则正向执行，否则反向执行
     * @param isPost 是否是通过postRedo或者postUndo执行的
     */
    private fun doOpt(bean: OptBean, isRedo: Boolean = true, isPost: Boolean = false) {
        when (bean.optType) {
            OptBean.OptType.ADD -> {
                if (isRedo) {
                    meaListener.getPointList().add(bean.pointAdd!!)
                } else {
                    meaListener.getPointList().removeAt(bean.addIndex!!)
//                    MeaUtil.removeLast(meaListener.getPointList(), bean.pointAdd!!)
                }
            }
            OptBean.OptType.MOVE -> {
                if (isRedo) {
                    meaListener.getPointList()[bean.moveIndex!!] = bean.pointMoveTo!!
                } else {
                    meaListener.getPointList()[bean.moveIndex!!] = bean.pointMoveFrom!!
                }
            }
            OptBean.OptType.INSERT -> {
                if (isRedo) {
                    meaListener.getPointList().add(bean.insertIndex!!, bean.pointInsert!!)
                } else {
                    meaListener.getPointList().removeAt(bean.insertIndex!!)
                }
            }
            OptBean.OptType.DELETE -> {
                if (isRedo) {
                    meaListener.getPointList().removeAt(bean.deleteIndex!!)
                } else {
                    meaListener.getPointList().add(bean.deleteIndex!!, bean.pointDelete!!)
                }
            }
            OptBean.OptType.TRNAS -> {
                val transLon = bean.pointTransEnd!!.x - bean.pointTransStart!!.x
                val transLat = bean.pointTransEnd!!.y - bean.pointTransStart!!.y
                if (isRedo) {
                    meaListener.getPointList().forEachIndexed { index, it ->
                        meaListener.getPointList()[index] = Point(
                            it.x + transLon,
                            it.y + transLat,
                            it.spatialReference
                        )
                    }
                } else {
                    meaListener.getPointList().forEachIndexed { index, it ->
                        meaListener.getPointList()[index] = Point(
                            it.x - transLon,
                            it.y - transLat,
                            it.spatialReference
                        )
                    }
                }
            }
        }
        meaListener.updateGraphic()
        if (isPost) {
            if (isRedo) {
                optRedoList.remove(bean)
                optUndoList.add(bean)
            } else {
                optRedoList.add(bean)
                optUndoList.remove(bean)
            }
        }
    }

    /**
     * 执行长按事件-弹出操作菜单
     */
    fun showMeaMenu(graphic: Graphic) {
        closeMarker()
        if (popMenuPoint != null) {
            val callout = meaListener.getMapView().callout
//            val view = LayoutInflater.from(context).inflate(R.layout.map_meamenu_layout, null, false)
//            val llMenu = view.findViewById<LinearLayout>(R.id.ll_mea_menu)
            val llMenu = LinearLayout(context).apply {
                layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT)
            }
            val funcList = arrayListOf<String>()
            if (graphic.geometry is Point) {
                funcList.add(MENU_DELETE)
            } else if (graphic.geometry is Polyline) {
                funcList.add(MENU_INSERT)
            } else if (graphic.geometry is Polygon) {
                if (meaListener.isTranslate()) {
                    funcList.add(MENU_STOPTRANS)
                } else {
                    funcList.add(MENU_TRANS)
                }
            }
            if (funcList.isNotEmpty()) {
                val params = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, MeaUtil.dp2px(context, 20f))
                params.gravity = Gravity.CENTER
                funcList.forEachIndexed { index, it ->
                    llMenu.addView(TextView(context).apply {
                        text = it
                        setTextColor(Color.WHITE)
                        textSize = 12f
                        gravity = Gravity.CENTER
                        minWidth = MeaUtil.dp2px(context, 30f)
                        layoutParams = params
                        tag = it
                        setOnClickListener(MenuClickListener())
                    })
                    if (funcList.size > 1 && index < funcList.size - 1) {
                        val divider = View(context)
                        divider.layoutParams = LinearLayout.LayoutParams(MeaUtil.dp2px(context, 1f), MeaUtil.dp2px(context, 20f))
                        divider.setBackgroundColor(Color.WHITE)
                        llMenu.addView(divider)
                    }
                }
                callout.refresh()
                meaListener.getMapView().callout.style.apply {
                    borderColor = ContextCompat.getColor(context, R.color.colorPrimary)
                    backgroundColor = ContextCompat.getColor(context, R.color.colorPrimary)
//                    minHeight = MeaUtil.dp2px(context, 30f)
                    minHeight = 5
                    cornerRadius = 5
                    leaderLength = 8
                }
                callout.show(llMenu, popMenuPoint)
            }
        }
    }

    /**
     * 关闭弹窗
     */
    fun closeMarker(): Boolean {
        if (meaListener.getMapView().callout.isShowing) {
            meaListener.getMapView().callout.dismiss()
            return true
        }
        return false
    }


    //菜单点击事件
    private inner class MenuClickListener : View.OnClickListener {
        override fun onClick(v: View) {
            when (v.tag.toString()) {
                "删除" -> if (funcIndex != -1) {
                    optDelete(popMenuPoint!!, funcIndex)
                }
                "插入" -> if (funcIndex != -1) {
                    optInsert(popMenuPoint!!, funcIndex)
                }
                "裁切" -> {
                }
                "平移" -> meaListener.startTranslate()
                "结束平移" -> meaListener.stopTranslate()
                "旋转" -> {
                }
            }
            closeMarker()
        }
    }
}